_blog/2023-08-30-using-oauth-with-axum.mdx

```diff
@@ -0,0 +1,440 @@
+    Extension(oauth_client): Extension<BasicClient>
+) -> Result<impl IntoResponse, impl IntoResponse> {
+
+    let token = match oauth_client
+        .exchange_code(AuthorizationCode::new(query.code))
+        .request_async(async_http_client)
+        .await {
+        Ok(res) => res,
+        Err(e) => {
+            error!("An error occurred while exchanging the code: {e}");
+            return Err((StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))
+            
+        }
+    };
+
+    let profile = match state.ctx.get("https://openidconnect.googleapis.com/v1/userinfo")
+        .bearer_auth(token.access_token().secret().to_owned())
+        .send().await {
+        Ok(res) => res,
+        Err(e) => return Err(ApiError::ReqwestError.to_error(e.to_string())),
+    };
+
+    let profile = profile.json::<UserProfile>().await.unwrap();
+
+    let secs: i64 = token.expires_in().unwrap().as_secs().try_into().unwrap();
+
+    let max_age = Local::now().naive_local() + Duration::seconds(secs);
+
+    let cookie = Cookie::build("sid", token.access_token().secret().to_owned())
+        .domain(".app.localhost")
+        .path("/")
+        .secure(true)
+        .http_only(true)
+        .max_age(TimeDuration::seconds(secs))
+        .finish();
+
+
+     if let Err(e) = sqlx::query("INSERT INTO users (email) VALUES ($1) ON CONFLICT (email) DO NOTHING")
+        .bind(profile.email.clone())
+        .execute(&state.db)
+        .await {
+        error!("Error while trying to make account: {e}")
+    }
+    
+    if let Err(e) = sqlx::query("INSERT INTO sessions (user_id, session_id, expires_at) VALUES (
+        (SELECT ID FROM USERS WHERE email = $1 LIMIT 1),
+         $2, $3)
+        ON CONFLICT (user_id) DO UPDATE SET 
+        session_id = excluded.session_id, 
+        expires_at = excluded.expires_at")
+        .bind(profile.email)
+        .bind(token.access_token().secret().to_owned())
+        .bind(max_age)
+        .execute(&state.db)
+        .await {
+        error!("Error while trying to make session: {e}");
+    }
+
+    Ok((
+        jar.add(cookie),
+        Redirect::to("/protected")
+    ))
+}
+```
+
+So now that we've finished handling the callback, we'll want to add our protected route. However, before that we need to make sure it's actually protected - we can do that by writing a middleware that will check whether or not the user has a cookie, then grabs the email from our own database, adds it as an extension to the request itself and then the protected route will return the extension we passed in.
+
+```rs
+pub async fn check_authenticated<B>(
+    State(state): State<AppState>,
+    jar: PrivateCookieJar,
+    mut req: Request<B>,
+    next: Next<B>
+) -> Result<impl IntoResponse, impl IntoResponse> {
+// return early if no cookie exists
+    let Some(cookie) = jar.get("sid").map(|cookie| cookie.value().to_owned()) else {
+        return Err((StatusCode::FORBIDDEN, "Forbidden!".to_string()));
```

Comment:
I believe these should be unauthorized: https://auth0.com/blog/forbidden-unauthorized-http-status-codes/

Reply:
fixed